### PR TITLE
Add X-API-KEY header authentication for access control

### PR DESCRIPTION
**Pull request by ruands**

### CODE REVIEW

**Comment by cobusc** [comment]
@ruands A lot of the files import `os`, but it is not used.
Also, can we define `API_KEY_HEADER = "X-API-KEY" in `access_control/settings.py` and then reference this in the tests and implementation? I saw one place where it is referenced as `X-Api-Key` while in all other places it seems to be `X-API-KEY`.

File: swagger_server/middleware.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import os
+from werkzeug.wrappers import Request, Response
+
+ALLOWED_API_KEYS = set(os.getenv("ALLOWED_API_KEYS").split(","))
```
Comment (cobusc) [inline swagger_server/middleware.py]: Can we move this to `access_control/settings.py`?

File: access_control/__init__.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import os
```
Comment (AltusBarry) [inline access_control/__init__.py]: Rather move this init and all tests to `access_control/tests/` just safer when it comes to altering env vars.

Comment (AltusBarry) [reply access_control/__init__.py]: I would personally even go so far as to add an if `if __name__ != "__main__":` as well.

File: access_control/settings.py [hunk 3]
```diff
@@ -5,3 +5,6 @@
     "DB_URI",
     "postgres+psycopg2://access_control:password@localhost:5432/access_control"
 )
+
+API_KEY_HEADER = "X-API-KEY"
+ALLOWED_API_KEYS = set(os.getenv("ALLOWED_API_KEYS").split(","))
```
Comment (AltusBarry) [inline access_control/settings.py]: My assumption is this is required, so don't use `getenv` use `os.environ["ALLOWED_API_KEYS"]`

File: swagger_server/middleware.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+class AuthMiddleware(object):
+
+    def __init__(self, app):
+        self.app = app
+
+    def __call__(self, environ, start_response):
+        request = Request(environ)
+        key = request.headers.get(API_KEY_HEADER, None)
+        if key:
```
Comment (AltusBarry) [inline swagger_server/middleware.py]: `if key and key in ALLOWED_API_KEYS:`

File: swagger_server/test/controllers/test_domain_controller.py [hunk 5]
```diff
@@ -4,10 +4,12 @@
 
 import random
 import uuid
+
```
Comment (AltusBarry) [inline swagger_server/test/controllers/test_domain_controller.py]:
```
python core
- openline
3rd party imports
- openline
framework imports
- openline
project imports
```

I'm guilty of not following this on this project as well. Could you check it just in this file for now please.

Comment (AltusBarry) [reply swagger_server/test/controllers/test_domain_controller.py]: Actually check it in all `access_control` files. Lets leave swagger_server for now.

File: swagger_server/test/controllers/test_domain_controller.py [hunk 6]
```diff
@@ -40,7 +44,8 @@ def test_domain_create(self):
             '/api/v1/domains',
             method='POST',
             data=json.dumps(data),
-            content_type='application/json')
+            content_type='application/json',
+            headers=self.headers)
```
Comment (AltusBarry) [inline swagger_server/test/controllers/test_domain_controller.py]: can't the client itself be altered in setUp?

File: swagger_server/test/test_exception_handlers.py [hunk 7]
```diff
@@ -2,14 +2,12 @@
...
 from __future__ import absolute_import
 
-import random
 import uuid
-import werkzeug
 
 from flask import json
 
-from swagger_server.models.domain import Domain  # noqa: E501
-from swagger_server.models.domain_update import DomainUpdate  # noqa: E501
```
Comment (AltusBarry) [inline swagger_server/test/test_exception_handlers.py]: Thanks for clearing out all these imports.

File: swagger_server/test/test_middleware.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
+import uuid
+
+import os
+
+from unittest.mock import patch
+
```
Comment (AltusBarry) [inline swagger_server/test/test_middleware.py]: Extra open line here, check import order.

File: swagger_server/test/test_middleware.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+            "name": ("%s" % uuid.uuid1())[:30],
+            "description": "a super cool test domain",
+        }
+        self.domain_model = db_actions.crud(
+            model="Domain",
+            api_model=Domain,
+            data=self.domain_data,
+            action="create"
+        )
+
```
Comment (AltusBarry) [inline swagger_server/test/test_middleware.py]: remove open line. Check if client can't rather be altered here.

Comment (AltusBarry) [reply swagger_server/test/test_middleware.py]: Ah no, these tests need to be tested without it in the header.

File: access_control/__init__.py (hunk 2 above)
Comment (ruands) [reply access_control/__init__.py]: I'm not entirely sure what to check here.

File: swagger_server/test/controllers/test_domain_controller.py (hunk 6 above)
Comment (ruands) [reply swagger_server/test/controllers/test_domain_controller.py]: Looked into the werkzeug docs and it seems the werkzeug client does not seem to allow it.

File: swagger_server/test/test_middleware.py (hunk 9 above)
Comment (ruands) [reply swagger_server/test/test_middleware.py]: Mentioned the setup issue earlier.

File: access_control/__init__.py (hunk 2 above)
Comment (AltusBarry) [reply access_control/__init__.py]:
```
if __name__ != "__main__":
    orig_environ = dict(os.environ)
    orig_environ["ALLOWED_API_KEYS"] = "test-api-key"
    os.environ.update(orig_environ) 
```

**Comment by cobusc** [comment]
:+1: And a special thanks for cleaning up the unused imports

**Pull request merged**